File: 208. Implement Trie (Prefix Tree).py
class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        :type word: str
        :rtype: None
        """
        n = len(word)
        node = self.root
        i = 0
        
        # go over each char in word
        while i < n:
            
            # if charNode != None:
                # this char exists in a branch already
                # directly go down a level to check the next char
            # else: # charNode == None
            
            # this char and remaining chars don't exist in a branch already
            if node.children[self.charToIndex(word[i])] == None:
                # insert a new trie node to represent the next char
                node.children[self.charToIndex(word[i])] = TrieNode()
            
           # go down a level
            node = node.children[self.charToIndex(word[i])] 
            i += 1 
                
        # label the end of word
        node.isEndOfWord = True
        

    def search(self, word):
        """
        :type word: str
        :rtype: bool
        """
        node = self.root
        n = len(word)
        i = 0
        while i < n:
            charNode = node.children[self.charToIndex(word[i])]
            # this char doesn't exist in trie
            if charNode == None:
                return False
            # this char exists. continue 
            else:
                node = charNode
                i += 1
        
        # all chars in word exist in trie
        
        # check if the word ends here in trie
        if node.isEndOfWord: 
            return True
        else:
            return False        
        

    def startsWith(self, prefix):
        """
        :type prefix: str
        :rtype: bool
        """
        node = self.root
        n = len(prefix)
        i = 0
        
        # go over each char in prefix
        while i < n:
            charNode = node.children[self.charToIndex(prefix[i])]
            # this char doesn't exist in trie
            if charNode == None:
                return False
            # this char exists. continue 
            else:
                node = charNode
                i += 1
        
        # all chars in prefix exist in trie
        return True
        
    def charToIndex(self, char):
        """
        :type char: str
        :rtype: int
        return the index of char in the list of children
        """
        return ord(char) - ord("a")
    

class TrieNode():
    
    def __init__(self, children = [None] * 26, isEndOfWord = False):
        self.children = children
        self.isEndOfWord = isEndOfWord


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 208. Implement Trie (Prefix Tree).java
class TrieNode {
    
    TrieNode[] children;
    Boolean isEndOfWord;        
    
    public TrieNode() {
        children = new TrieNode[26];
        isEndOfWord = false;            
    }
}

class Trie {
    
    TrieNode root;
    
    public Trie() {
        root = new TrieNode();
    }
    
    public void insert(String word) {
        int n = word.length();
        
        // pointers
        TrieNode node = root;
        int i = 0;
        
        // go over each char in word
        while (i < n) {
            // this char and remaining chars don't exist in a branch already
            if (node.children[charToIndex(word.charAt(i))] == null) {
                // insert a new trie node to represent the next char
                node.children[charToIndex(word.charAt(i))] = new TrieNode();
            } // else: this char already exists in a branch. go down a level directly to check the next char      
            // go down a level
            node = node.children[charToIndex(word.charAt(i))];
            i += 1;
        }
        
        // label the end of word
        node.isEndOfWord = true;
        
    }
    
    public boolean search(String word) {
        int n = word.length();
        // pointers
        TrieNode node = root;
        int i = 0;
        while (i < n) {
            TrieNode charNode = node.children[charToIndex(word.charAt(i))];
            // this char doesn't exist in trie
            if (charNode == null) {
                return false;
            // this char exists. continue to check the next char
            } else {
                node = charNode;
                i += 1;                    
            }
        }
        // all chars in word exist in trie
        
        // check if the word ends here in trie
        if (node.isEndOfWord) {
            return true;
        } else {
            return false;
        }
        
    }
    
    public boolean startsWith(String prefix) {
        int n = prefix.length();
        // pointers
        TrieNode node = root;
        int i = 0;
        while (i < n) {
            TrieNode charNode = node.children[charToIndex(prefix.charAt(i))];
            // this char doesn't exist in trie
            if (charNode == null) {
                return false;
            // this char exists. continue to check the next char
            } else {
                node = charNode;
                i += 1;                    
            }
        }
        // all chars in prefix exist in trie  
        return true;
    }
    
    public int charToIndex(Character c) {
        return c - 'a';
    }
}



/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */